File: cpp/1249.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<int> st;
        vector<int> valid, invalid;
    
        for(int i=0;i<s.length();i++){
            if(s[i]=='('){
                st.push(i);
            } else if(s[i]==')'){
                if(st.empty()) invalid.push_back(i);
                else {
                    valid.push_back(st.top());
                    valid.push_back(i);
                    st.pop();
                }
            }
        }
        sort(valid.begin(),valid.end());
        int ind=0,indIn=0;
        int szValid = valid.size(), szInvalid = invalid.size();
        string ans="";
        for(int i=0;i<s.length();i++){
            if(s[i]!='(' && s[i]!=')')ans+=s[i];
            else {
                if(indIn<szInvalid && invalid[indIn]==i){
                    indIn++;
                    continue;
                }
                if(ind<szValid && valid[ind]==i){
                    ans+=s[i];
                    ind++;
                }
            }
        }
        return ans;
    }
};

File: golang/1249.go
type IntStack struct {
    ar []int
}

func (st *IntStack) push (ele int){
    st.ar = append(st.ar, ele)
}

func (st *IntStack) size () int {
    return len(st.ar)
}

func (st *IntStack) pop() int {
    topEle := st.ar[st.size()-1]
    st.ar = st.ar[:st.size()-1]
    return topEle
}

func minRemoveToMakeValid(s string) string {
    runeAr := []rune(s)
    var openBrSt IntStack
    for index, ele := range runeAr{
        if ele == '(' {
            openBrSt.push(index)
        } else if ele==')'{
            if openBrSt.size() == 0 {
                runeAr[index] = '#'
            } else {
                openBrSt.pop()
            }
        }
    }
    
    for ;openBrSt.size()>0; {
        ind := openBrSt.pop()
        runeAr[ind] = '#'
    }
    
    ans := ""
    for _, ele := range runeAr {
        if ele != '#' {
            ans += string(ele)
        }
    }
    return ans
}
